src/main/scala/org/camunda/feel/impl/interpreter/FeelInterpreter.scala

```diff
@@ -945,16 +945,16 @@ class FeelInterpreter {
         values + (name -> v) // zip
   }
 
-  private def filterList(list: List[Val], filter: Val => Val): Val =
-    list match {
-      case Nil => ValList(List())
-      case x :: xs =>
-        withBoolean(filter(x), _ match {
-          case false => filterList(xs, filter)
-          case true =>
-            withList(filterList(xs, filter), l => ValList(x :: l.items))
-        })
-    }
+  private def filterList(list: List[Val], filter: Val => Val): Val = {
+    val conditionNotFulfilled = ValString("_")
+
+    mapEither[Val, Val](list,
+      item => withBoolean(filter(item), {
+        case true => item
+        case false => conditionNotFulfilled
+      }).toEither,
+      items => ValList(items.filterNot(_ == conditionNotFulfilled)))
+  }
```

Comment:
It would be nice to use `filter()` but we need to handle also the failure cases. If a filter doesn't evaluate to a boolean then we want to report the failure. 

I'm not sure how to combine this failure handling with a `filter()`. I'm open to suggestions.

Reply:
Ah I understand it now. Evaluation errors result in `ValError` which is also a `Val` and turns into a left with the `.toEither`. It makes sense to me now.